MM-11692 Add infinite scroll to webapp

sudheerDev opened the pull request:
#### Summary
Adding infinite scroll to webapp.

#### Ticket Link
[MM-11692](https://mattermost.atlassian.net/browse/MM-11692)
[MM-11504](https://mattermost.atlassian.net/browse/MM-11504)

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [ ] Ran `make check-style` to check for style errors (required for all pull requests)
- [ ] Ran `make test` to ensure unit and component tests passed
- [ ] Added or updated unit tests (required for all new features)
- [ ] Needs to be implemented in mobile (link to PR or User Story)
- [ ] Has server changes (please link)
- [ ] Has redux changes (please link)
- [ ] Has UI changes
- [ ] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-webapp/blob/master/i18n/en.json)) updates
- [ ] Touches critical sections of the codebase (auth, posting, etc.)

enahum commented:
@sudheerDev how early are you loading more posts?

sudheerDev commented:
@enahum https://github.com/mattermost/mattermost-webapp/pull/2032/commits/5a70b94042215b708e457888f222eb3bd711708a#diff-1f1a548021bbfb77959c436de8fb935aR458

600px but it is arbitrary in a way. I would like to have it tested and see tweak based on it.

enahum commented:
When I say tests is for loading posts when scrolling

esethna commented:
@sudheerDev @enahum will a test server work on this PR? given it's not against master?

sudheerDev commented:
No i don't think so. It needs a different server branch. We need to have it set manually or have it merged to post-metadata branch so, we can update existing spinmint to test this.

mattermod commented:
Spinmint test server created at: https://i-07dd9ed641d5e0e28.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-07dd9ed641d5e0e28

enahum commented:
Manually

sudheerDev commented:
> is there a way you can add tests to this

I will get back to this tomorrow. I have an existing ticket in this sprint to add tests and it was intended mainly for this tests.

asaadmahmood commented:
@esethna @sudheerDev The height given currently is `150`, and padding `60`, we should remove the padding and only have a height of `40`.

esethna commented:
@sudheerDev please help make the adjustments above proposed by AM?

sudheerDev commented:
@esethna Made changes 
2. changed heights based on AM's suggestions

1. Started loading posts a little earlier now. Have a look at the new changes once.
    Start Loading when content reaches at 50% of screen height. i.e if 800px height of postList it now starts to load when content left above fold is 400px. We decreased loader height and i removed an requestAnimationFrame so these should also help.

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0d9c97efe285e7f82.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-0d9c97efe285e7f82

sudheerDev commented:
Needed to add the atBottom flag back as i saw a case when image posted by user will not correct scroll position as `this.checkBottom()` is for present UI state which will have image below the fold so it never correct.

hmhealey commented on components/post_view/post_list.jsx:
```diff
@@ -618,6 +619,19 @@ export default class PostList extends React.PureComponent {
+            topRow = (
+                <div
+                    className='loading-screen'
+                    style={{position: 'relative', height: '40px', padding: '0px'}}
```
Can this be moved into the CSS?

sudheerDev replied:
Done

hmhealey commented on components/post_view/post_list.jsx:
```diff
@@ -347,43 +333,45 @@ export default class PostList extends React.PureComponent {
-                element.scrollIntoView();
-            }
+        if (this.resizeAnimationFrame) {
+            window.cancelAnimationFrame(this.resizeAnimationFrame);
```
Also, we're not clearing the value of `this.resizeAnimationFrame` when the callback is fired so it will always try to cancel the previous callback even if it's already been called

sudheerDev replied:
Though the function says resize it was being used to call for any postList scroll change actions(Just the way it was). 

We don't need to clear the value for cancelling the animationFrame. It is similar to timeout the value exists if there is a pending animation frame.

hmhealey replied:
Since it's in an if statement, I would expect it to only call `cancelAnimationFrame` some of the time, but because it doesn't clear the value of `this.resizeAnimationFrame`, it will "cancel" it basically every time. It works fine and there's nothing majorly wrong with it, but the code is written in a way that's a bit confusing

mattermod commented:
Spinmint test server created at: https://i-0cb1c1d966b4717ce.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-0cb1c1d966b4717ce

sudheerDev commented on components/post_view/post_list.jsx:
```diff
@@ -252,51 +241,49 @@ export default class PostList extends React.PureComponent {
-            if (this.previousScrollHeight !== postList.scrollHeight && posts[0].id === prevPosts[0].id) {
-                postList.scrollTop = this.previousScrollTop + (postList.scrollHeight - this.previousScrollHeight);
+            if (snapshot.previousScrollHeight !== postlistScrollHeight && posts[0].id === prevPosts[0].id) {
+                window.requestAnimationFrame(() => {
```
Chrome does not correct it well if we scroll fast to the top and cause few extra calls before correcting properly. 
hence the `requestAnimationFrame`. I would have preferred to avoid but i couldn't figure another way and happens only on chrome so i guess it has something to do with chrome scroll optimisations. I only guess is to check this once in a while and remove when chrome corrects it fine.

sudheerDev commented on components/post_view/post_list.jsx:
```diff
@@ -252,51 +241,49 @@ export default class PostList extends React.PureComponent {
         if (this.extraPagesLoaded > MAX_EXTRA_PAGES_LOADED) {
             // Prevent this from loading a lot of pages in a channel with only hidden messages
+            // Enable load more messages manual link
+            this.state({autoRetryEnable: false});
```
Aah thanks nice catch. Would have made a mess 👍

sudheerDev commented on components/post_view/post_list.jsx:
```diff
@@ -347,43 +333,45 @@ export default class PostList extends React.PureComponent {
-            this.previousScrollHeight = postList.scrollHeight;
-            this.previousScrollTop = postList.scrollTop;
-            this.previousClientHeight = postList.clientHeight;
+        this.resizeAnimationFrame = window.requestAnimationFrame(() => {
```
This gets called quite a few times if there is not postmeta data and it will stop the scroll because of JS thread lock so adding animation frame request so as to let user continue his actions.

sudheerDev replied:
About this yes, I think it is better to have this as i don't see any downsides here and it helps in fixing couple of odd cases where we are generating odd scroll corrections. 
I just noticed that reaction list in webapp triggers scrollCorrect as redux selector returns a new object if it is an empty list. 

This logic helps not to lock Javascript thread in cases around like this.

hmhealey commented:
This should be merged into master once the post-metadata branch has been merged

sudheerDev commented:
@hmhealey We should be merging this to post-metadata branch and have dylan test it

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

hmhealey commented:
I'm regenerating the test server now that the serverside changes for post metadata have been merged

mattermod commented:
Spinmint test server created at: https://i-00aa1a0131a8e79b7.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-00aa1a0131a8e79b7

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-03d9b432b7d25f5bb.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-03d9b432b7d25f5bb

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0c3643cf6fb4e3d1d.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-0c3643cf6fb4e3d1d

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-02f864a18cce01d23.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-02f864a18cce01d23

DHaussermann commented:
This looks good to me. Markdown images are now working.
- Tested 3 different formats for markdown image
- Tested post with mark down created by webhooks
- Scroll is still infinite
- Also checked RN with markdown images

mattermod commented:
Spinmint test running for more than 7 days. This test server was terminated.

mattermod commented:
Spinmint test running for more than 7 days. This test server was terminated.

rgarrigue commented:
Any update on this one ?
